lib/ble.toit

```diff
@@ -257,23 +270,157 @@ class Client:
+    return characteristics_[uuid]
+
+
+BLE_CHR_TYPE_READ_ONLY_    ::= 1
+BLE_CHR_TYPE_WRITE_ONLY_   ::= 2
+BLE_CHR_TYPE_READ_WRITE_   ::= 3
+BLE_CHR_TYPE_NOTIFICATION_ ::= 4
+
+BLE_WAIT_RECV_             ::= 1 << 0
+BLE_WAIT_ACCESSED_         ::= 1 << 1
+BLE_WAIT_SUBSCRIBED_       ::= 1 << 2
+
+abstract class Characteristic:
+  /**
+  The UUID of the characteristic.
+  */
+  uuid/uuid_pkg.Uuid
+
+  state_/ResourceState_
+
+  constructor service/Service resource .uuid:
+    state_ = ResourceState_ service.resource_group_ resource
+  
+/** 
+A characteristic that can only be read by clients.
+*/
+class ReadOnlyCharacteristic extends Characteristic:
+  value_/ByteArray := #[]
+
+  constructor service/Service uuid/uuid_pkg.Uuid value:  
+    resource := ble_add_server_characteristic_ service.resource_ uuid.to_byte_array BLE_CHR_TYPE_READ_ONLY_ value
+    super service resource uuid
+    value_ = value
+
+  value -> ByteArray: return value_
+
+  value= value/ByteArray -> none:
+    ble_set_characteristics_value_ state_.resource value
+    value_ = value
+
+/**
+A characteristic that can only be written to by clients.
+*/
+class WriteOnlyCharacteristic extends Characteristic:
+  constructor service/Service uuid/uuid_pkg.Uuid:
+    resource := ble_add_server_characteristic_ service.resource_ uuid.to_byte_array BLE_CHR_TYPE_WRITE_ONLY_ null
+    super service resource uuid
+
+  value -> ByteArray?:
+    state_.wait_for_state BLE_WAIT_RECV_
+    data := ble_get_characteristics_value_ state_.resource
+    state_.clear_state BLE_WAIT_RECV_
+    return data;
+
+/**
+A characteristic that allows both read and write by the client.
+*/
+class ReadWriteCharacteristic extends Characteristic:
+  constructor service/Service uuid/uuid_pkg.Uuid value:  
+    resource := ble_add_server_characteristic_ service.resource_ uuid.to_byte_array BLE_CHR_TYPE_READ_WRITE_ value
+    super service resource uuid
+
+  value -> ByteArray?:
+    state_.wait_for_state BLE_WAIT_RECV_
+    data := ble_get_characteristics_value_ state_.resource
+    state_.clear_state BLE_WAIT_RECV_
+    return data;
+
+  value= value/ByteArray -> none:
+    ble_set_characteristics_value_ state_.resource value
+
+
+/** 
+A characteristic that the client can subscribe to changes on.
+*/
+class NotificationCharacteristic extends Characteristic:
+  constructor service/Service uuid/uuid_pkg.Uuid:  
+    resource := ble_add_server_characteristic_ service.resource_ uuid.to_byte_array BLE_CHR_TYPE_NOTIFICATION_ null
+    super service resource uuid
+
+  value= value/ByteArray -> none:
+    ble_notify_characteristics_value_ state_.resource value
+
 /**
 The local BLE device.
 
 After construction, the BLE stack is initialized and ready to use.
+
+If services is not empty, sets up the services for the advertiser. 
 */
 class Device:
   resource_group_ := ?
   resource_state_/monitor.ResourceState_? := null
 
-  constructor.default:
-    resource_group_ = ble_init_
+  server_config/ServerConfig?
```

Comment:
Maybe this could just be called `server`? It looks like you're expecting folks to write code that manipulates this after the device has been created (like `device.server_config.add_xxx`), so I guess this is intentionally public?

Reply:
Removed the field. The server config is freezed on ble create time